scalalib/src/mill/scalalib/ScalaModule.scala

```diff
@@ -23,7 +23,19 @@ trait ScalaModule extends mill.Module with TaskModule { outer =>
     override def moduleDeps = Seq(outer)
   }
   def scalaVersion: T[String]
-  def mainClass: T[Option[String]] = None
+
+  def mainClass: T[Option[String]] = T{
+    discoverMainClasses() match {
+      case Seq(main) => Some(main)
+      case _         => None
+    }
+  }
+
+  def discoverMainClasses: T[Seq[String]] = T{
+    Task.traverse(transitiveModuleDeps){ module => T.task {
```

Comment:
I believe SBT does not discover main classes in upstream modules, only in the current one. If we don't have any strong opinion in how it should be, let's follow their behavior